pkg/lightning/backend/local/region_job.go

```diff
@@ -501,9 +503,23 @@ func (local *Backend) doWrite(ctx context.Context, j *regionJob) error {
 	//nolint: errcheck
 	defer iter.Close()
 
-	var remainingStartKey []byte
+	// init to start and remainingStartKey, to found duplication across multiple engines
+	var (
+		remainingStartKey []byte
+		needDupCheck      bool
+		lastKey4DupCheck  []byte
+	)
+	if _, ok := j.ingestData.(*external.MemoryIngestData); ok {
+		needDupCheck = true
+	}
 	for iter.First(); iter.Valid(); iter.Next() {
 		k, v := iter.Key(), iter.Value()
+		if needDupCheck {
+			if lastKey4DupCheck != nil && bytes.Equal(lastKey4DupCheck, k) {
+				return errors.Errorf("duplicate key found: %s", hex.EncodeToString(lastKey4DupCheck))
+			}
+			lastKey4DupCheck = k
+		}
```

Comment:
maybe add in `loadBatchRegionData` where there is a sort phase, more simpler to check

Reply:
After the sorting of `loadBatchRegionData` there's no place to iterate all keys, so I need to add an iteration. I think it's better for performance to use current iteration here.

Reply:
seems no performance penalty, we already do a byte compare `bytes.Compare(e.memKVsAndBuffers.keys[i], e.memKVsAndBuffers.keys[k]) < 0`, but I'm not sure, sorty's less function have 4 params

Reply:
sorty do concurrent sorting in multiple goroutine, but we will need a single thread iteration and check

Reply:
I'll check the guarantee of sorty. Not sure if it will split [1,2,3,3,4,5] into [1,2,3] and [3,4,5], the duplicate value is split into 2 sorting group so the sorting function can't find duplication

Reply:
i guess `sorty` still need to compare or merge those 2 parts, else how would it keep the whole slice sorted?